File: Heap/Minimum Cost to Connect Sticks/ConnectSticks.cpp
class ConnectSticks{
public:
    int connectSticks(vector<int>& arr){
        //min heap to store the sticks
        priority_queue<int, <vector<int>, greater<int>> pq;

        //add all sticks to the queue
        for(int stick : arr){
            pq.push(stick);
        }

        //track the total cost
        int totCost = 0;

        //combine sticks while there's more than one stick left
        while(pq.size() > 1){
            //get the two smallest sticks
            int s1 = pq.top();
            pq.pop();

            int s2 = pq.top();
            pq.pop();

            //get the cost of adding these sticks by adding the sticks together
            int cost - s1 + s1;

            //add that cost to the total cost
            totCost += cost;

            //push back the new stick to the queue
            pq.push(cost);
        }

        return totCost; //return the total cost
    }

};

File: Heap/Minimum Cost to Connect Sticks/ConnectSticks.java
import java.util.PriorityQueue;

public class ConnectSticks {
    public int connectSticks(int[] arr){
        //priority queue to get the smallest sticks first
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        //add all the sticks to the queue
        for(int stick : arr){
            pq.add(stick);
        }

        int totCost = 0; //total cost of connecting the sticks

        //until one stick left, combine two smallest sticks
        while(pq.size() > 1){
            //take out two smallest sticks
            int s1 = pq.remove();
            int s2 = pq.remove();

            //cost to connect two sticks
            int cost = s1 + s2;

            //add to the total cost
            totCost += cost;

            //add the new stick back to the queue
            pq.add(cost);
        }

        return totCost;
    }

    public static void main(String[] args){
        ConnectSticks cs = new ConnectSticks();

        int[] sticks = {1,5,9,11,12,50,30,2,6};
        int totalCost = cs.connectSticks(sticks);

        System.out.println(totalCost);
    }
}
